File: android/app/src/main/java/com/smartcardpoc/CprScannerModule.java
package com.smartcardpoc;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactPackage;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ViewManager;
import java.util.Arrays;
import java.util.List;
import java.util.Collections;
import java.util.ArrayList;

public class CprScannerModule extends ReactContextBaseJavaModule {
    private CprScanner cprScanner;

    public CprScannerModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.cprScanner = new CprScanner(reactContext.getBaseContext(), reactContext);
    }

    @Override
    public String getName() {
        return "CprScanner";
    }

    @ReactMethod
    public void initialize(Promise promise) {
        promise.resolve("Initialized");
    }

    @ReactMethod
    public void resume() {
        if (cprScanner != null) {
            cprScanner.resume();
        }
    }

    @ReactMethod
    public void pause() {
        if (cprScanner != null) {
            cprScanner.pause();
        }
    }

    @Override
    public void invalidate() {
        super.invalidate();
        if (cprScanner != null) {
            cprScanner.destroy();
            cprScanner = null;
        }
    }
}

class CprScannerPackage implements ReactPackage {
    @Override
    public List<NativeModule> createNativeModules(ReactApplicationContext reactContext) {
        List<NativeModule> modules = new ArrayList<>();
        modules.add(new CprScannerModule(reactContext));
        return modules;
    }

    @Override
    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
        return Collections.emptyList();
    }
}

File: android/app/src/main/java/com/smartcardpoc/CprScannerModule.kt
package com.smartcardpoc

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.LifecycleEventListener
import android.util.Log

class CprScannerModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener {

    private val reactContext: ReactApplicationContext = reactContext
    private var cprScanner: CprScanner? = null
    private var isInitialized = false

    init {
        Log.d("CprScannerModule", "Module initialized")
        reactContext.addLifecycleEventListener(this)
    }

    override fun getName(): String = "CprScanner"

    @ReactMethod
    fun startScanning() {
        Log.d("CprScannerModule", "startScanning called")
        cprScanner?.resume()
    }

    @ReactMethod
    fun stopScanning() {
        Log.d("CprScannerModule", "stopScanning called")
        cprScanner?.pause()
    }

    override fun onHostResume() {
        Log.d("CprScannerModule", "onHostResume")
        try {
            if (!isInitialized) {
                cprScanner = CprScanner(reactContext) // Instantiate here
                               isInitialized = true
            }
        } catch (e: Exception) {
            Log.e("CprScannerModule", "Error in onHostResume: ${e.message}", e)
        }
    }

    override fun onHostPause() {
        Log.d("CprScannerModule", "onHostPause")
        cprScanner?.pause()
    }

    override fun onHostDestroy() {
        Log.d("CprScannerModule", "onHostDestroy")
        cprScanner?.pause()
        cprScanner = null
        reactContext.removeLifecycleEventListener(this)
    }
}